File: instrumentation/httpurlconnection/library/src/main/java/io/opentelemetry/instrumentation/library/httpurlconnection/internal/HttpUrlHttpAttributesGetter.java
/*
 * Copyright The OpenTelemetry Authors
 * SPDX-License-Identifier: Apache-2.0
 */

package io.opentelemetry.instrumentation.library.httpurlconnection.internal;

import static java.util.Collections.emptyList;
import static java.util.Collections.singletonList;

import io.opentelemetry.instrumentation.api.semconv.http.HttpClientAttributesGetter;
import java.net.HttpURLConnection;
import java.net.URLConnection;
import java.util.List;
import javax.annotation.Nullable;

public class HttpUrlHttpAttributesGetter
        implements HttpClientAttributesGetter<URLConnection, Integer> {

    @Override
    public String getHttpRequestMethod(URLConnection connection) {
        HttpURLConnection httpURLConnection = (HttpURLConnection) connection;
        return httpURLConnection.getRequestMethod();
    }

    @Override
    public String getUrlFull(URLConnection connection) {
        return connection.getURL().toExternalForm();
    }

    @Override
    public List<String> getHttpRequestHeader(URLConnection connection, String name) {
        String value = connection.getRequestProperty(name);
        return value == null ? emptyList() : singletonList(value);
    }

    @Override
    public Integer getHttpResponseStatusCode(
            URLConnection connection, Integer statusCode, @Nullable Throwable error) {
        return statusCode;
    }

    @Override
    public List<String> getHttpResponseHeader(
            URLConnection connection, Integer statusCode, String name) {
        String value = connection.getHeaderField(name);
        return value == null ? emptyList() : singletonList(value);
    }

    @Nullable
    @Override
    public String getNetworkProtocolName(URLConnection connection, @Nullable Integer integer) {
        // HttpURLConnection hardcodes the protocol name&version
        return "http";
    }

    @Nullable
    @Override
    public String getNetworkProtocolVersion(URLConnection connection, @Nullable Integer integer) {
        // HttpURLConnection hardcodes the protocol name&version
        return "1.1";
    }

    @Override
    public String getServerAddress(URLConnection connection) {
        return connection.getURL().getHost();
    }

    @Override
    public Integer getServerPort(URLConnection connection) {
        return connection.getURL().getPort();
    }
}

File: instrumentation/httpurlconnection/library/src/main/java/io/opentelemetry/instrumentation/library/httpurlconnection/internal/HttpUrlHttpAttributesGetter.kt
/*
 * Copyright The OpenTelemetry Authors
 * SPDX-License-Identifier: Apache-2.0
 */

package io.opentelemetry.instrumentation.library.httpurlconnection.internal

import io.opentelemetry.instrumentation.api.semconv.http.HttpClientAttributesGetter
import java.net.HttpURLConnection
import java.net.URLConnection

internal class HttpUrlHttpAttributesGetter : HttpClientAttributesGetter<URLConnection, Int> {
    override fun getHttpRequestMethod(connection: URLConnection): String {
        val httpURLConnection = connection as HttpURLConnection
        return httpURLConnection.requestMethod
    }

    override fun getUrlFull(connection: URLConnection): String = connection.getURL().toExternalForm()

    override fun getHttpRequestHeader(
        connection: URLConnection,
        name: String,
    ): MutableList<String?> =
        when (val value = connection.getRequestProperty(name)) {
            null -> mutableListOf()
            else -> mutableListOf(value)
        }

    override fun getHttpResponseStatusCode(
        connection: URLConnection,
        statusCode: Int,
        error: Throwable?,
    ): Int = statusCode

    override fun getHttpResponseHeader(
        connection: URLConnection,
        statusCode: Int,
        name: String,
    ): MutableList<String?> =
        when (val value = connection.getHeaderField(name)) {
            null -> mutableListOf()
            else -> mutableListOf(value)
        }

    // HttpURLConnection hardcodes the protocol name&version
    override fun getNetworkProtocolName(
        connection: URLConnection,
        integer: Int?,
    ): String = "http"

    // HttpURLConnection hardcodes the protocol name&version
    override fun getNetworkProtocolVersion(
        connection: URLConnection,
        integer: Int?,
    ): String = "1.1"

    override fun getServerAddress(connection: URLConnection): String = connection.getURL().host

    override fun getServerPort(connection: URLConnection): Int = connection.getURL().port
}
